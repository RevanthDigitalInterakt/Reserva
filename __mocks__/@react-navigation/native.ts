const navigate = jest.fn();

const useNavigation = () => ({
  navigate,
});

export {
  useNavigation,
};
