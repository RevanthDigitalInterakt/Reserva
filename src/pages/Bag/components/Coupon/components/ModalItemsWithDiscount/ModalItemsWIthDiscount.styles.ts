import { StyleSheet } from 'react-native';
import { COLORS } from '../../../../../../base/styles/colors';

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    backgroundColor: COLORS.MODAL_BACKGROUND_COLOR,
    justifyContent: 'flex-end',
  },

  modalContent: {
    height: '60%',
    backgroundColor: COLORS.WHITE,
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    paddingHorizontal: 10,
  },

  modalTitleContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 20,
    alignItems: 'center',
  },

  modalTitle: {
    fontFamily: 'ReservaSerif-Bold',
    fontSize: 24,
  },

  modalRow: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  modalSubTitle: {
    fontFamily: 'ReservaSans-Regular',
  },

  modalDivider: {
    borderBottomWidth: 1,
    borderBottomColor: COLORS.DIVIDER,
    marginVertical: 20,
  },

  modalItemContainer: {
    borderWidth: 1,
    borderColor: COLORS.DIVIDER,
    paddingHorizontal: 15,
    marginVertical: 5,
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },

  modalItemTitle: {
    color: COLORS.BLACK,
    fontFamily: 'ReservaSans-Bold',
    textTransform: 'uppercase',
    margin: 10,
    padding: 10,
  },
});

export default styles;
