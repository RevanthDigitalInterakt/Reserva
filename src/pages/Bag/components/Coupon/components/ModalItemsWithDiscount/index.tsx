import React from 'react';
import {
  Modal,
  Text,
  View,
  TouchableOpacity,
  FlatList,
} from 'react-native';
import { useBagCouponItemsWithDiscount } from '../../../../../../zustand/useBagCouponItemsWithDiscount/useBagCouponItemsWithDiscount';
import IconClose from '../../../../../../../assets/icons/IconClose';

export default function ModalItemsWithDiscount() {
  const { showModalItems, setShowModalItems, productItems } = useBagCouponItemsWithDiscount(['showModalItems', 'setShowModalItems', 'productItems']);

  return (
    <Modal animationType="slide" transparent visible={showModalItems}>
      <View
        style={{
          flex: 1,
          backgroundColor: 'rgba(0, 0, 0, 0.5)',
          justifyContent: 'flex-end',
        }}
      >
        <View
          style={{
            height: '60%',
            backgroundColor: 'white',
            borderTopLeftRadius: 10,
            borderTopRightRadius: 10,
            paddingHorizontal: 10,
          }}
        >
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              marginVertical: 20,
              alignItems: 'center',
            }}
          >
            <Text style={{ fontFamily: 'ReservaSerif-Bold', fontSize: 24 }}>Descontos do Pedido</Text>
            <TouchableOpacity onPress={() => setShowModalItems(false)}>
              <IconClose />
            </TouchableOpacity>
          </View>

          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Text style={{ fontFamily: 'ReservaSans-Regular', fontSize: 16 }}>O desconto </Text>
            <Text style={{ fontFamily: 'ReservaSans-Regular', fontWeight: 'bold', fontSize: 14 }}>foi aplicado </Text>
            <Text style={{ fontFamily: 'ReservaSans-Regular', fontSize: 16 }}>aos seguintes itens:</Text>
          </View>

          <View style={{ borderBottomWidth: 1, borderBottomColor: '#d7d7d7', marginVertical: 20 }} />

          <FlatList
            data={productItems}
            keyExtractor={(item) => item.id}
            showsVerticalScrollIndicator={false}
            renderItem={({ item }) => (
              <View
                style={{
                  borderWidth: 1,
                  borderColor: '#D7D7D7',
                  paddingHorizontal: 15,
                  marginVertical: 5,
                  flexDirection: 'row',
                  alignItems: 'center',
                  flex: 1,
                }}
              >
                <View
                  style={{
                    width: 35,
                    height: 35,
                    borderRadius: 1,
                    backgroundColor: '#e5e5e5',
                  }}
                />
                <Text
                  numberOfLines={1}
                  style={{
                    color: '#000000',
                    fontFamily: 'ReservaSans-Bold',
                    textTransform: 'uppercase',
                    margin: 10,
                    padding: 10,
                  }}
                >
                  {item.productName}
                </Text>
              </View>
            )}
          />
        </View>
      </View>
    </Modal>
  );
}
