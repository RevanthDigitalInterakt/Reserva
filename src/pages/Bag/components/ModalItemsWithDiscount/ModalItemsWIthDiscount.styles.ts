import { StyleSheet } from 'react-native';
import { COLORS } from '../../../../base/styles/colors';
import { FONTS } from '../../../../base/styles';

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    backgroundColor: COLORS.MODAL_BACKGROUND_COLOR,
    justifyContent: 'flex-end',
  },

  modalContent: {
    height: '65%',
    backgroundColor: COLORS.WHITE,
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    paddingHorizontal: 10,
  },

  modalTitleContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 20,
    alignItems: 'center',
  },

  modalTitle: {
    fontFamily: FONTS.RESERVA_SERIF_BOLD,
    fontSize: 24,
  },

  modalRow: {
    flexDirection: 'row',
  },

  modalSubTitle: {
    fontFamily: FONTS.RESERVA_SANS_REGULAR,
  },

  modalDivider: {
    borderBottomWidth: 1,
    borderBottomColor: COLORS.DIVIDER,
    marginVertical: 20,
  },

  modalItemContainer: {
    borderWidth: 1,
    borderColor: COLORS.DIVIDER,
    paddingHorizontal: 15,
    marginVertical: 5,
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },

  modalItemMedia: {
    width: 40,
    height: 40,
  },

  modalItemTitle: {
    color: COLORS.BLACK,
    fontFamily: FONTS.RESERVA_SANS_BOLD,
    textTransform: 'uppercase',
    margin: 10,
    padding: 10,
  },

  modalFooter: {
    padding: 10,
  },

  modalFooterTextInfo: {
    color: COLORS.INPUT_ERROR_MESSAGE,
    marginTop: -2,
    marginLeft: 5,
    fontFamily: FONTS.RESERVA_SANS_MEDIUM,
    fontSize: 14,
  },

  modalFooterRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 10,
  },

  modalFooterTitle: {
    fontFamily: FONTS.RESERVA_SANS_REGULAR,
    fontSize: 14,
  },

  modalFooterLabel: {
    fontFamily: FONTS.RESERVA_SANS_BOLD,
    fontSize: 16,
  },

  modalFooterLabelAlignment: {
    alignItems: 'flex-end',
  },
});

export default styles;
