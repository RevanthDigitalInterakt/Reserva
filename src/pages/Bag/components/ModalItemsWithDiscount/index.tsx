import React, { useCallback } from 'react';
import {
  Modal,
  Text,
  View,
  TouchableOpacity,
  FlatList,
  Image,
} from 'react-native';
import { useBagCouponItemsWithDiscount } from '../../../../zustand/useBagCouponItemsWithDiscount/useBagCouponItemsWithDiscount';
import IconClose from '../../../../../assets/icons/IconClose';
import styles from './ModalItemsWIthDiscount.styles';
import IconInfoFill from '../../../../../assets/icons/IconInfoFill';
import { COLORS } from '../../../../base/styles/colors';
import { useBagStore } from '../../../../zustand/useBagStore/useBagStore';

export type TItemData = {
  id: string;
  name: string;
  sellingPrice: number;
  itemColor: string;
  imageUrl?: string | null | undefined;
  imageSource: string;
};

export default function ModalItemsWithDiscount() {
  const { showModalItems, setShowModalItems } = useBagCouponItemsWithDiscount(['showModalItems', 'setShowModalItems']);

  const { marketingData, appTotalizers } = useBagStore(['marketingData', 'appTotalizers']);

  const splitDecimalNumber = (value: number) => (`${value?.toFixed(2)}`)?.split('.')[1];

  const splitIntegerNumber = (numInteger: number) => (numInteger <= 0
    ? Math.ceil(numInteger)
    : Math.floor(numInteger));

  const renderItemList = (item: TItemData) => (
    <View style={styles.modalItemContainer}>
      <Image
        style={styles.modalItemMedia}
        source={{ uri: item.imageSource }}
        resizeMode="contain"
      />
      <Text
        numberOfLines={1}
        style={styles.modalItemTitle}
      >
        {item.name}
      </Text>
    </View>
  );

  const totalPriceItemsWithDiscount = useCallback(() => {
    const newValue = marketingData?.itemsWithCouponDiscount
      .reduce((acc, cur) => acc + cur.sellingPrice, 0);

    if (newValue) {
      const formattValue = newValue / 100;

      const decimalNumber = splitDecimalNumber(formattValue);
      const integerNumber = splitIntegerNumber(formattValue);

      return {
        decimalNumber,
        integerNumber,
      };
    }

    return {
      decimalNumber: 0,
      integerNumber: 0,
    };
  }, [marketingData?.itemsWithCouponDiscount]);

  return (
    <Modal animationType="slide" transparent visible={showModalItems}>
      <View style={styles.modalContainer}>
        <View style={styles.modalContent}>

          <View style={styles.modalTitleContainer}>
            <Text style={styles.modalTitle}>Descontos do Pedido</Text>
            <TouchableOpacity onPress={() => setShowModalItems(false)}>
              <IconClose />
            </TouchableOpacity>
          </View>

          <View style={styles.modalRow}>
            <Text style={[styles.modalSubTitle, { fontSize: 16 }]}>O desconto </Text>
            <Text style={[styles.modalSubTitle, { fontWeight: 'bold', fontSize: 14 }]}>foi aplicado </Text>
            <Text style={[styles.modalSubTitle, { fontSize: 16 }]}>aos seguintes itens:</Text>
          </View>

          <View style={styles.modalDivider} />

          <FlatList
            testID="com.usereserva:id/list_items_with_discount"
            data={marketingData?.itemsWithCouponDiscount}
            keyExtractor={(item) => item.id}
            showsVerticalScrollIndicator={false}
            renderItem={({ item }) => renderItemList(item)}
          />

          <View style={styles.modalFooter}>
            <View style={styles.modalRow}>
              <IconInfoFill />
              <Text style={styles.modalFooterTextInfo}>
                Os demais itens do carrinho já possuem um desconto
                maior ou não são elegíveis para este desconto!
              </Text>
            </View>

            <View style={styles.modalDivider} />

            <View style={styles.modalFooterRow}>
              <View>
                <Text style={styles.modalFooterTitle}>
                  Produtos elegíveis
                </Text>
                <View style={styles.modalRow}>
                  <Text style={[styles.modalFooterLabel, {
                    textDecorationLine: 'line-through',
                  }]}
                  >
                    R$
                    {' '}
                    {totalPriceItemsWithDiscount().integerNumber}
                    ,
                  </Text>
                  <View style={{ alignSelf: 'flex-start' }}>
                    <Text
                      style={
                        [
                          styles.modalFooterLabel,
                          { fontSize: 10, textDecorationLine: 'line-through' },
                        ]
                      }
                    >
                      {totalPriceItemsWithDiscount().decimalNumber}
                    </Text>
                  </View>
                </View>
              </View>
              <View>
                <Text style={styles.modalFooterTitle}>
                  Descontos de cupom
                </Text>
                <View style={styles.modalFooterLabelAlignment}>
                  <View style={styles.modalRow}>
                    <Text style={[styles.modalFooterLabel, {
                      color: COLORS.SWITCH_ACTIVE,
                    }]}
                    >
                      R$
                      {' '}
                      {splitIntegerNumber(appTotalizers.discount).toString().replace('-', '')}
                      ,
                    </Text>
                    <View style={{ alignSelf: 'flex-start' }}>
                      <Text
                        style={
                          [
                            styles.modalFooterLabel,
                            { fontSize: 10, color: COLORS.SWITCH_ACTIVE },
                          ]
                        }
                      >
                        {splitDecimalNumber(appTotalizers.discount)}
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </Modal>
  );
}
