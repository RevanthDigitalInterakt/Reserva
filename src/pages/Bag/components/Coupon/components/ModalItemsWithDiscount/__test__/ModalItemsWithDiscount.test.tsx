import React from 'react';
import { theme } from '@usereservaapp/reserva-ui';
import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components/native';
import { render, screen } from '@testing-library/react-native';
import CartContextProvider from '../../../../../../../context/CartContext';
import ModalItemsWithDiscount from '..';

const Component = (
  <ThemeProvider theme={theme}>
    <MockedProvider addTypename={false}>
      <CartContextProvider>
        <ModalItemsWithDiscount />
      </CartContextProvider>
    </MockedProvider>
  </ThemeProvider>
);

describe('Modal items with discount', () => {
  beforeEach(() => {
    jest.useFakeTimers({ legacyFakeTimers: true });
  });
  it('snapshot', () => {
    render(Component);

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('render items in flatlist', () => {
    const { getByTestId } = render(Component);

    expect(getByTestId('com.usereserva:id/list_items_with_discount'));
  });
});
