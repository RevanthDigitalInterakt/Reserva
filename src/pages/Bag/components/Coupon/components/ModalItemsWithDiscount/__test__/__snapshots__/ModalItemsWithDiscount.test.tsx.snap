// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal items with discount snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  transparent={true}
  visible={false}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderTopLeftRadius": 10,
          "borderTopRightRadius": 10,
          "height": "65%",
          "paddingHorizontal": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontFamily": "ReservaSerif-Bold",
              "fontSize": 24,
            }
          }
        >
          Descontos do Pedido
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight="24"
            bbWidth="24"
            fill="none"
            focusable={false}
            height="24"
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 24,
                  "width": 24,
                },
              ]
            }
            vbHeight={24}
            vbWidth={24}
            width="24"
          >
            <RNSVGGroup
              fill={null}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGGroup>
                <RNSVGPath
                  d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
                  fill={4278190080}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "ReservaSans-Regular",
              },
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          O desconto 
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "ReservaSans-Regular",
              },
              Object {
                "fontSize": 14,
                "fontWeight": "bold",
              },
            ]
          }
        >
          foi aplicado 
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "ReservaSans-Regular",
              },
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          aos seguintes itens:
        </Text>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "#d7d7d7",
            "borderBottomWidth": 1,
            "marginVertical": 20,
          }
        }
      />
      <RCTScrollView
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={Array []}
        testID="com.usereserva:id/list_items_with_discount"
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
      <View
        style={
          Object {
            "padding": 10,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight="16"
            bbWidth="16"
            fill="none"
            focusable={false}
            height="16"
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 16,
                  "width": 16,
                },
              ]
            }
            vbHeight={16}
            vbWidth={16}
            width="16"
          >
            <RNSVGGroup
              fill={null}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGGroup
                clipPath="clip0_468_6202"
              >
                <RNSVGCircle
                  cx={7.66}
                  cy={7.66}
                  fill={4293132331}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                  r={6.66}
                />
              </RNSVGGroup>
              <RNSVGPath
                d="M7 4.33301H8.33333V5.66634H7V4.33301ZM7 6.99967H8.33333V10.9997H7V6.99967Z"
                fill={4294967295}
                propList={
                  Array [
                    "fill",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
          <Text
            style={
              Object {
                "color": "#E4002B",
                "fontFamily": "ReservaSans-Medium",
                "fontSize": 14,
                "marginLeft": 5,
                "marginTop": -2,
              }
            }
          >
            Os demais itens do carrinho já possuem um desconto maior ou não são elegíveis para este desconto!
          </Text>
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#d7d7d7",
              "borderBottomWidth": 1,
              "marginVertical": 20,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <View>
            <Text
              style={
                Object {
                  "fontFamily": "ReservaSans-Regular",
                  "fontSize": 14,
                }
              }
            >
              Produtos elegíveis
            </Text>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "ReservaSans-Bold",
                      "fontSize": 16,
                    },
                    Object {
                      "textDecorationLine": "line-through",
                    },
                  ]
                }
              >
                R$
                 
                0
                ,
              </Text>
              <View
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "ReservaSans-Bold",
                        "fontSize": 16,
                      },
                      Object {
                        "fontSize": 10,
                        "textDecorationLine": "line-through",
                      },
                    ]
                  }
                >
                  0
                </Text>
              </View>
            </View>
          </View>
          <View>
            <Text
              style={
                Object {
                  "fontFamily": "ReservaSans-Regular",
                  "fontSize": 14,
                }
              }
            >
              Descontos de cupom
            </Text>
            <View
              style={
                Object {
                  "alignItems": "flex-end",
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "ReservaSans-Bold",
                        "fontSize": 16,
                      },
                      Object {
                        "color": "#31B94F",
                      },
                    ]
                  }
                >
                  R$
                   
                  0
                  ,
                </Text>
                <View
                  style={
                    Object {
                      "alignSelf": "flex-start",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "ReservaSans-Bold",
                          "fontSize": 16,
                        },
                        Object {
                          "color": "#31B94F",
                          "fontSize": 10,
                        },
                      ]
                    }
                  >
                    00
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
